File: NEsper/eql/subquery/SubqueryJoinedPropDesc.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

namespace net.esper.eql.subquery
{
    /// <summary>
    /// Holds property information for joined properties in a subquery.
    /// </summary>
    public class SubqueryJoinedPropDesc
    {
        private String indexPropName;
        private Type coercionType;
        private String keyPropName;
        private int? keyStreamId;

        /// <summary>Ctor.</summary>
        /// <param name="indexPropName">is the property name of the indexed field</param>
        /// <param name="coercionType">is the type to coerce to</param>
        /// <param name="keyPropName">is the property name of the key field</param>
        /// <param name="keyStreamId">is the stream number of the key field</param>
        public SubqueryJoinedPropDesc(String indexPropName, Type coercionType, String keyPropName, int? keyStreamId)
        {
            this.indexPropName = indexPropName;
            this.coercionType = coercionType;
            this.keyPropName = keyPropName;
            this.keyStreamId = keyStreamId;
        }

        /// <summary>Returns the property name of the indexed field.</summary>
        /// <returns>property name of indexed field</returns>
        public String IndexPropName
        {
            get { return indexPropName; }
        }

        /// <summary>Returns the coercion type of key to index field.</summary>
        /// <returns>type to coerce to</returns>
        public Type CoercionType
        {
            get { return coercionType; }
        }

        /// <summary>Returns the property name of the key field.</summary>
        /// <returns>property name of key field</returns>
        public String KeyPropName
        {
            get { return keyPropName; }
        }

        /// <summary>Returns the stream id of the key field.</summary>
        /// <returns>stream id</returns>
        public int? KeyStreamId
        {
            get { return keyStreamId; }
        }

        /// <summary>Returns the key stream numbers.</summary>
        /// <param name="descList">a list of descriptors</param>
        /// <returns>key stream numbers</returns>
        public static int[] GetKeyStreamNums(Collection<SubqueryJoinedPropDesc> descList)
        {
            int[] streamIds = new int[descList.Size()];
            int count = 0;
            foreach (SubqueryJoinedPropDesc desc in descList)
            {
                streamIds[count++] = desc.GetKeyStreamId();
            }
            return streamIds;
        }

        /// <summary>Returns the key property names.</summary>
        /// <param name="descList">a list of descriptors</param>
        /// <returns>key property names</returns>
        public static String[] GetKeyProperties(ICollection<SubqueryJoinedPropDesc> descList)
        {
            String[] result = new String[descList.Size()];
            int count = 0;
            foreach (SubqueryJoinedPropDesc desc in descList)
            {
                result[count++] = desc.GetKeyPropName();
            }
            return result;
        }

        /// <summary>Returns the key coercion types.</summary>
        /// <param name="descList">a list of descriptors</param>
        /// <returns>key coercion types</returns>
        public static Type[] GetCoercionTypes(ICollection<SubqueryJoinedPropDesc> descList)
        {
            Type[] result = new Class[descList.Size()];
            int count = 0;
            foreach (SubqueryJoinedPropDesc desc in descList)
            {
                result[count++] = desc.GetCoercionType();
            }
            return result;
        }
    }
}

File: trunk/esper/src/main/java/net/esper/eql/subquery/SubqueryJoinedPropDesc.java
package net.esper.eql.subquery;

import java.util.Collection;

/**
 * Holds property information for joined properties in a subquery.
 */
public class SubqueryJoinedPropDesc
{
    private String indexPropName;
    private Class coercionType;
    private String keyPropName;
    private Integer keyStreamId;

    /**
     * Ctor.
     * @param indexPropName is the property name of the indexed field
     * @param coercionType is the type to coerce to
     * @param keyPropName is the property name of the key field
     * @param keyStreamId is the stream number of the key field
     */
    public SubqueryJoinedPropDesc(String indexPropName, Class coercionType, String keyPropName, Integer keyStreamId)
    {
        this.indexPropName = indexPropName;
        this.coercionType = coercionType;
        this.keyPropName = keyPropName;
        this.keyStreamId = keyStreamId;
    }

    /**
     * Returns the property name of the indexed field.
     * @return property name of indexed field
     */
    public String getIndexPropName()
    {
        return indexPropName;
    }

    /**
     * Returns the coercion type of key to index field.
     * @return type to coerce to
     */
    public Class getCoercionType()
    {
        return coercionType;
    }

    /**
     * Returns the property name of the key field.
     * @return property name of key field
     */
    public String getKeyPropName()
    {
        return keyPropName;
    }

    /**
     * Returns the stream id of the key field.
     * @return stream id
     */
    public Integer getKeyStreamId()
    {
        return keyStreamId;
    }

    /**
     * Returns the key stream numbers.
     * @param descList a list of descriptors
     * @return key stream numbers
     */
    public static int[] getKeyStreamNums(Collection<SubqueryJoinedPropDesc> descList)
    {
        int[] streamIds = new int[descList.size()];
        int count = 0;
        for (SubqueryJoinedPropDesc desc : descList)
        {
            streamIds[count++] = desc.getKeyStreamId();
        }
        return streamIds;
    }

    /**
     * Returns the key property names.
     * @param descList a list of descriptors
     * @return key property names
     */
    public static String[] getKeyProperties(Collection<SubqueryJoinedPropDesc> descList)
    {
        String[] result = new String[descList.size()];
        int count = 0;
        for (SubqueryJoinedPropDesc desc : descList)
        {
            result[count++] = desc.getKeyPropName();
        }
        return result;
    }

    /**
     * Returns the key coercion types.
     * @param descList a list of descriptors
     * @return key coercion types
     */
    public static Class[] getCoercionTypes(Collection<SubqueryJoinedPropDesc> descList)
    {
        Class[] result = new Class[descList.size()];
        int count = 0;
        for (SubqueryJoinedPropDesc desc : descList)
        {
            result[count++] = desc.getCoercionType();
        }
        return result;
    }
}
